apiml-common/src/main/java/org/zowe/apiml/product/version/VersionService.java

```diff
@@ -0,0 +1,123 @@
+    private String zoweManifest;
+
+    public VersionService() {
+        buildInfo = new BuildInfo().getBuildInfoDetails();
+    }
+
+    public VersionService(BuildInfoDetails buildInfo) {
+        this.buildInfo = buildInfo;
+    }
+
+    /**
+     * Getting the cached VersionInfo object, if it's empty it will be filled
+     * @return filled VersionInfo object
+     */
+    public VersionInfo getVersion() {
+        if (version.getApimlVersion() == null) {
+            updateVersionInfo();
+        }
+        return version;
+    }
+
+    /**
+     * Updating the cached VersionInfo object with values from Zowe's manifest.json, API ML's build-info.properties
+     * and git.properties files
+     */
+    public void updateVersionInfo() {
+        if (StringUtils.isNotEmpty(zoweManifest)) {
+            version.setZoweVersion(getZoweVersion(zoweManifest));
+        }
+        version.setApimlVersion(getApimlVersion());
+    }
+
+    /**
+     * Retrieving the information about API ML version from build-info.properties and git.properties files
+     * @return the version, build and commit numbers in one string
+     */
+    private String getApimlVersion() {
+        String apimlVersion = NO_VERSION;
+        if (!buildInfo.getVersion().equalsIgnoreCase("unknown")) {
+            apimlVersion = String.format("%s build #%s (%s)", buildInfo.getVersion(), buildInfo.getNumber(), buildInfo.getCommitId());
+        }
+        return apimlVersion;
+    }
+
+    /**
+     * Retrieving the information about Zowe version from manifest.json file
+     * @param manifestJsonFile the path to Zowe's manifest.json file
+     * @return the version and build numbers in one string
+     */
+    private String getZoweVersion(String manifestJsonFile) {
+        ObjectMapper mapper = new ObjectMapper();
+        mapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);
+        try {
+            String manifestJson = FileUtils.readFile(manifestJsonFile);
+            if (manifestJson != null) {
+                ObjectNode objectNode = mapper.readValue(manifestJson, ObjectNode.class);
+                JsonNode versionNode = objectNode.get("version");
+                if (versionNode != null && !versionNode.asText().isEmpty()) {
+                    StringBuilder zoweVersion = new StringBuilder();
+                    zoweVersion.append(versionNode.asText());
+                    String buildNumber = "n/a";
+                    JsonNode buildNode = objectNode.get("build");
+                    if (buildNode != null) {
+                        JsonNode buildNumberNode = buildNode.get("number");
+                        if (buildNumberNode != null && StringUtils.isNotEmpty(buildNumberNode.asText())) {
+                            buildNumber = buildNumberNode.asText();
+                        }
+                    }
+                    zoweVersion.append(" build #");
+                    zoweVersion.append(buildNumber);
```

Comment:
Would it be possible to move at least the internal logic to the separate private method to make this method more easily understood?

Reply:
Sure! Even into two private methods!